# COSMIC database preprocessing

This script uses the cosmic VCF file (Coding or Non coding) and Mutant export data (tsv) to generate tumor site counts per variant.

#### Software requirements:

- Python v3.8 or higher
- progressbar2 (python package; `pip install progressbar2`)
- sh (python package; `pip install sh`)
- Mac OSX (>=10.11) and Linux (Kernel version 5.x)

Hardware requirements:

- 2 CPU cores
- 6-8 GB RAM

Commandline arguments

```bash
python process_cosmic_data.py CosmicCodingMuts.vcf.gz CosmicMutantExport.tsv.gz
```

- CosmicCodingMut.vcf.gz - VCF file of all coding mutations in the current release.
- CosmicMutantExport.tsv.gz - A tab separated table of all COSMIC coding point mutations from targeted and genome wide screens from the current release.
- Further details at https://cancer.sanger.ac.uk/cosmic/download

By default, output file `cosmic_processed.txt.gz` is generated by the script.
Output file structure example:

```
11      66204714        C       T       COSM6715269;COSM6715270 large_intestine=1
```

#### Fields:

1. Chromosome
2. Position
3. Ref
4. Alt
5. COSMIC ID(s). Multiple COSMIC IDs are delimited by `;`
6. Frequency distribution across tumor types. Multiple entries are delimited by `;`

##### _NOTE: input files must be gzipped (by default gzipped with downloaded from COSMIC's database)_

### _<span style="color:red">This repo does **NOT** provide access to COSMIC database or license information. Users should obtain their own license to use the COSMIC database.</span>_
